dataprochub/proxy.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Extends default Proxy to limit which routes are added to the proxy. """
+
+import re
+from jupyterhub.proxy import ConfigurableHTTPProxy
+
+class RedirectProxy(ConfigurableHTTPProxy):
+  """ Extends ConfigurableHTTPProxy to handle redirects instead of routing. """
+
+  def __init__(self, **kwargs):
+    self.log.info('# Initializing RedirectProxy.')
+    super().__init__(**kwargs)
+
+  async def add_route(self, routespec, target, data):
+    """ Adds route only if target is not a Component Gateway URL.
+
+    Using the default add_route would return a 404 when accessing /user/<USER>.
+    """
+    reg = r'(https:\/\/)*[a-zA-Z0-9]*-dot-[a-z1-9-]*\.dataproc\.googleusercontent\.com'
```

Comment:
Could this also be more specific? For instance, maybe we should only block this for `/user` routespecs that target Component Gateway.

Reply:
I chose the Component Gateway in case later versions of JupyterHub adds other routes for /user/... that are not just the server and that we might need.

Reply:
Yeah I mean we should check that it's a /user/<foo> routespec and a Component Gateway URL to block it.

Reply:
Gotcha

Reply:
Fixed in 7d23da939c1e101356da7a7df3871d20c47d2422